File: backend/src/benchmark.ts
/**
 * Reproducible benchmark runner (spec §25, §26).
 *
 * Layers:
 *   text      — Layer A: standalone snippets, one finalized turn each.
 *   streaming — Layer B: event sequences with an expected state after every event.
 *   audio     — Layer C: requires recorded synthetic audio + live transcription;
 *               reported as skipped (never fabricated) when prerequisites are absent.
 *
 * Gold labels are read-only inputs. Results are written to
 * backend/data/benchmark_results.json and a UI copy of the streaming summary to
 * src/data/streaming_benchmark_results.json. Real numbers only.
 */
import { existsSync, readFileSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

import { BACKEND_DIR, getSettings } from './config.ts';
import { DeterministicExtractor, EXTRACTOR_VERSION } from './deterministicExtractor.ts';
import { DuplicateEventError, EncounterService } from './encounterService.ts';
import { EvidenceIndex } from './evidenceIndex.ts';
import { Speaker, activeWarnings, newId } from './models.ts';
import { NO_MATCH_PRIMARY, NO_MATCH_SECONDARY } from './warningEngine.ts';

const REPO_DIR = path.resolve(BACKEND_DIR, '..');

const ALLOWED_MESSAGES = new Set([NO_MATCH_PRIMARY, NO_MATCH_SECONDARY]);

function round4(x: number): number {
  return Math.round(x * 10000) / 10000;
}

function safeDiv(a: number, b: number): number {
  return b === 0 ? 0.0 : a / b;
}

function median(sorted: number[]): number {
  const n = sorted.length;
  if (n % 2 === 1) return sorted[(n - 1) / 2];
  return (sorted[n / 2 - 1] + sorted[n / 2]) / 2;
}

export function buildService(): EncounterService {
  const settings = getSettings();
  const index = new EvidenceIndex(settings.evidence_path, settings.synonym_path, {
    strict: true,
    allowPendingVerification: settings.evidence_allow_pending_verification,
  });
  return new EncounterService(settings, index, new DeterministicExtractor(index));
}

// ---------------------------------------------------------------------------
// Layer A — text
// ---------------------------------------------------------------------------

interface TextCase {
  id: string;
  category: string;
  input: string;
  expectedResultState: string;
  expectedEvidenceRecordId: string | null;
  expectedAbstention: boolean;
  expectedMedication?: string | null;
  expectedMedicationStatus?: string | null;
  [key: string]: unknown;
}

export async function runTextLayer(service: EncounterService): Promise<Record<string, unknown>> {
  const cases = (
    JSON.parse(readFileSync(path.join(BACKEND_DIR, 'data', 'benchmark_cases.json'), 'utf8')) as {
      cases: TextCase[];
    }
  ).cases;
  let tp = 0, fp = 0, fn = 0, tn = 0;
  let retrievalCorrect = 0;
  let abstExpected = 0, abstCorrect = 0;
  let negTotal = 0, negCorrect = 0, histTotal = 0, histCorrect = 0;
  let expectedPositive = 0, citedCorrect = 0;
  let unsupported = 0;
  let statusTotal = 0, statusCorrect = 0;
  let passCount = 0;
  const results: Array<Record<string, unknown>> = [];

  for (const testCase of cases) {
    const runtime = service.createEncounter();
    const snapshot = await service.processFinalTurn(runtime, {
      event_id: newId('evt'),
      text: testCase.input,
      speaker: Speaker.DOCTOR,
    });
    const actualState = snapshot.result_state;
    const active = activeWarnings(snapshot);
    const actualRecord =
      active.length === 1
        ? active[0].evidence_record_id
        : active.length === 0
          ? null
          : [...active.map((w) => w.evidence_record_id)].sort().join(',');
    const stateOk = actualState === testCase.expectedResultState;
    const recordOk = actualRecord === testCase.expectedEvidenceRecordId;
    const passed = stateOk && recordOk;
    if (passed) passCount++;

    const expPos = testCase.expectedResultState === 'EVIDENCE_FOUND';
    const actPos = actualState === 'EVIDENCE_FOUND';
    if (expPos) expectedPositive++;
    if (actPos && expPos && recordOk) tp++;
    else if (actPos && (!expPos || !recordOk)) fp++;
    if (expPos && (!actPos || !recordOk)) fn++;
    if (!actPos && !expPos) tn++;
    if (recordOk) retrievalCorrect++;
    if (testCase.expectedAbstention) {
      abstExpected++;
      if (!actPos) abstCorrect++;
    }
    if (testCase.category === 'explicit_negation') {
      negTotal++;
      if (stateOk) negCorrect++;
    }
    if (testCase.category === 'historical_use') {
      histTotal++;
      if (stateOk) histCorrect++;
    }
    if (actPos) {
      const record = service.index.getRecord(active[0].evidence_record_id);
      if (recordOk && String(record.sourceUrl).trim() && String(record.sourceSection).trim()) {
        citedCorrect++;
      }
    }
    for (const message of snapshot.messages) {
      if (!ALLOWED_MESSAGES.has(message)) unsupported++;
    }
    // Medication status accuracy for the interacting medication, when labeled.
    const expectedStatus = testCase.expectedMedicationStatus;
    if (expectedStatus && testCase.expectedMedication) {
      statusTotal++;
      const interacting = snapshot.assertions.filter(
        (a) =>
          a.category === 'other_medication' && a.concept_id in service.index.ontology.medication_concepts,
      );
      if (interacting.some((a) => a.status === expectedStatus)) statusCorrect++;
    }

    results.push({
      id: testCase.id,
      category: testCase.category,
      input: testCase.input,
      expectedResultState: testCase.expectedResultState,
      actualResultState: actualState,
      expectedEvidenceRecordId: testCase.expectedEvidenceRecordId,
      actualEvidenceRecordId: actualRecord,
      pass: passed,
      lookupReason: snapshot.lookup_reason,
    });
  }

  const precision = safeDiv(tp, tp + fp);
  const recall = safeDiv(tp, tp + fn);
  return {
    layer: 'A-text',
    caseCount: cases.length,
    metrics: {
      passRate: round4(safeDiv(passCount, cases.length)),
      triggerPrecision: round4(precision),
      triggerRecall: round4(recall),
      triggerF1: round4(safeDiv(2 * precision * recall, precision + recall)),
      retrievalAccuracy: round4(safeDiv(retrievalCorrect, cases.length)),
      correctAbstentionRate: round4(safeDiv(abstCorrect, abstExpected)),
      negationAccuracy: round4(safeDiv(negCorrect, negTotal)),
      historicalContextAccuracy: round4(safeDiv(histCorrect, histTotal)),
      citationCoverage: round4(safeDiv(citedCorrect, expectedPositive)),
      medicationStatusAccuracy: round4(safeDiv(statusCorrect, statusTotal)),
      falsePositiveCount: fp,
      unsupportedClaimCount: unsupported,
    },
    confusion: { truePositive: tp, falsePositive: fp, falseNegative: fn, trueNegative: tn },
    counts: { passCount, total: cases.length },
    cases: results,
  };
}

// ---------------------------------------------------------------------------
// Layer B — streaming
// ---------------------------------------------------------------------------

interface StreamingEvent {
  type: string;
  sequence?: number;
  text?: string;
  speaker?: string;
  event_id?: string;
  medication_surface_text?: string;
  cancelsEventOfSequence?: number;
  isDuplicateReplay?: boolean;
  expectedResultStates: string[];
  expectedEvidenceRecordIds: string[];
  expectedActiveWarningCount: number;
  expectRetractionWithReason?: boolean;
  expectedWarningContext?: string;
}

interface StreamingCase {
  id: string;
  category: string;
  events: StreamingEvent[];
}

export async function runStreamingLayer(service: EncounterService): Promise<Record<string, unknown>> {
  const data = JSON.parse(
    readFileSync(path.join(BACKEND_DIR, 'data', 'streaming_benchmark_cases.json'), 'utf8'),
  ) as { cases: StreamingCase[] };
  const cases = data.cases;
  let eventTotal = 0, eventCorrect = 0;
  let finalCorrect = 0;
  let retrievalEvents = 0, retrievalCorrect = 0;
  let retractionExpected = 0, retractionCorrect = 0;
  let prematureWarnings = 0;
  let duplicateWarnings = 0;
  let citationEvents = 0, citationCorrect = 0;
  let unsupported = 0;
  const retractionLatencyTurns: number[] = [];
  const caseResults: Array<Record<string, unknown>> = [];

  for (const testCase of cases) {
    const runtime = service.createEncounter();
    const proposalBySequence = new Map<number, string>();
    const eventResults: Array<Record<string, unknown>> = [];
    const warningCreatedAtEvent = new Map<string, number>();
    for (let eventIndex = 0; eventIndex < testCase.events.length; eventIndex++) {
      const event = testCase.events[eventIndex];
      try {
        if (event.type === 'transcript.final') {
          await service.processFinalTurn(runtime, {
            event_id: event.event_id || newId('evt'),
            text: event.text ?? '',
            speaker: event.speaker as Speaker,
            sequence: event.sequence,
          });
        } else if (event.type === 'prescription.proposed') {
          const snapshot = await service.proposePrescription(runtime, {
            event_id: newId('evt'),
            surface_text: event.medication_surface_text ?? '',
          });
          proposalBySequence.set(
            event.sequence as number,
            snapshot.proposals[snapshot.proposals.length - 1].proposal_id,
          );
        } else if (event.type === 'prescription.cancelled') {
          const ref = event.cancelsEventOfSequence as number;
          await service.cancelPrescription(runtime, {
            event_id: newId('evt'),
            proposal_id: proposalBySequence.get(ref) as string,
          });
        }
      } catch (err) {
        if (!(err instanceof DuplicateEventError) || !event.isDuplicateReplay) throw err;
      }

      const snapshot = runtime.snapshot;
      const active = activeWarnings(snapshot);
      const activeRecords = [...new Set(active.map((w) => w.evidence_record_id))].sort();
      const stateOk = event.expectedResultStates.includes(snapshot.result_state);
      const expectedSorted = [...event.expectedEvidenceRecordIds].sort();
      const recordsOk =
        activeRecords.length === expectedSorted.length &&
        activeRecords.every((r, i) => r === expectedSorted[i]);
      const countOk = active.length === event.expectedActiveWarningCount;
      let retractionOk = true;
      if (event.expectRetractionWithReason) {
        retractionExpected++;
        const retracted = snapshot.warnings.filter((w) => w.state === 'retracted');
        retractionOk = retracted.length > 0 && retracted.every((w) => Boolean(w.retraction_reason));
        if (retractionOk) retractionCorrect++;
        for (const w of retracted) {
          const createdAt = warningCreatedAtEvent.get(w.warning_id);
          if (createdAt !== undefined) {
            retractionLatencyTurns.push(eventIndex - createdAt);
            warningCreatedAtEvent.delete(w.warning_id);
          }
        }
      }
      let contextOk = true;
      if (event.expectedWarningContext) {
        contextOk = active.length > 0 && active.every((w) => w.context === event.expectedWarningContext);
      }
      for (const w of active) {
        if (!warningCreatedAtEvent.has(w.warning_id)) {
          warningCreatedAtEvent.set(w.warning_id, eventIndex);
        }
      }
      if (event.expectedEvidenceRecordIds.length) {
        retrievalEvents++;
        if (recordsOk) retrievalCorrect++;
        citationEvents += active.length;
        for (const w of active) {
          const record = service.index.getRecord(w.evidence_record_id);
          if (String(record.sourceUrl).trim() && String(record.sourceSection).trim()) {
            citationCorrect++;
          }
        }
      }
      if (event.expectedActiveWarningCount === 0 && active.length) {
        prematureWarnings += active.length;
      }
      if (active.length > event.expectedActiveWarningCount) {
        duplicateWarnings += active.length - event.expectedActiveWarningCount;
      }
      for (const message of snapshot.messages) {
        if (!ALLOWED_MESSAGES.has(message)) unsupported++;
      }

      const ok = stateOk && recordsOk && countOk && retractionOk && contextOk;
      eventTotal++;
      if (ok) eventCorrect++;
      eventResults.push({
        sequence: event.sequence ?? null,
        type: event.type,
        expectedResultStates: event.expectedResultStates,
        actualResultState: snapshot.result_state,
        expectedEvidenceRecordIds: event.expectedEvidenceRecordIds,
        actualEvidenceRecordIds: activeRecords,
        pass: ok,
      });
    }
    const finalOk = eventResults[eventResults.length - 1].pass as boolean;
    if (finalOk) finalCorrect++;
    caseResults.push({
      id: testCase.id,
      category: testCase.category,
      pass: eventResults.every((e) => e.pass),
      events: eventResults,
    });
  }

  const latencies: number[] = [];
  for (const runtime of service.encounters.values()) {
    for (const latency of runtime.latencies) {
      latencies.push(latency.total_ms);
    }
  }
  let latencySummary: Record<string, unknown> | null = null;
  if (latencies.length) {
    const sorted = [...latencies].sort((a, b) => a - b);
    latencySummary = {
      samples: sorted.length,
      medianMs: round4(median(sorted)),
      p90Ms: round4(sorted[Math.floor(0.9 * (sorted.length - 1))]),
      maxMs: round4(sorted[sorted.length - 1]),
      note: 'Backend pipeline only (final transcript received -> result computed), deterministic extractor, no network. Live-mode model latency is not included and must be measured separately.',
    };
  }

  return {
    layer: 'B-streaming',
    caseCount: cases.length,
    metrics: {
      casePassRate: round4(safeDiv(caseResults.filter((c) => c.pass).length, cases.length)),
      perEventStateAccuracy: round4(safeDiv(eventCorrect, eventTotal)),
      finalStateAccuracy: round4(safeDiv(finalCorrect, cases.length)),
      retrievalAccuracy: round4(safeDiv(retrievalCorrect, retrievalEvents)),
      warningRetractionAccuracy: round4(safeDiv(retractionCorrect, retractionExpected)),
      meanRetractionLatencyEvents: retractionLatencyTurns.length
        ? round4(safeDiv(retractionLatencyTurns.reduce((a, b) => a + b, 0), retractionLatencyTurns.length))
        : 0.0,
      prematureWarningCount: prematureWarnings,
      duplicateWarningCount: duplicateWarnings,
      citationCoverage: round4(safeDiv(citationCorrect, citationEvents)),
      unsupportedClaimCount: unsupported,
    },
    counts: { events: eventTotal, eventsCorrect: eventCorrect },
    processingLatency: latencySummary,
    cases: caseResults,
  };
}

// ---------------------------------------------------------------------------
// Layer C — audio (prerequisite-gated; never fabricated)
// ---------------------------------------------------------------------------

export function runAudioLayer(): Record<string, unknown> {
  const manifest = JSON.parse(
    readFileSync(path.join(BACKEND_DIR, 'data', 'audio_benchmark_manifest.json'), 'utf8'),
  ) as { cases: Array<{ file: string }> };
  const settings = getSettings();
  const missingFiles = manifest.cases
    .map((c) => c.file)
    .filter((file) => !existsSync(path.join(BACKEND_DIR, 'data', file)));
  const reasons: string[] = [];
  if (missingFiles.length) {
    reasons.push(`${missingFiles.length} audio files not recorded yet`);
  }
  if (!settings.openai_api_key) {
    reasons.push('no OPENAI_API_KEY for live transcription');
  }
  if (reasons.length) {
    return {
      layer: 'C-audio',
      status: 'SKIPPED',
      reason: reasons.join('; '),
      caseCount: manifest.cases.length,
      note: 'Results are only reported for runs that actually happened. Record the manifest audio and set OPENAI_API_KEY to execute this layer.',
    };
  }
  return {
    layer: 'C-audio',
    status: 'NOT_IMPLEMENTED_IN_THIS_RUNNER_VERSION',
    note: 'Transcribe each file through the realtime pipeline and evaluate against the manifest gold labels.',
  };
}

// ---------------------------------------------------------------------------

export async function main(layers: string[]): Promise<void> {
  const service = buildService();
  const output: Record<string, any> = {
    metadata: {
      benchmarkVersion: '0.3.0',
      evidenceVersion: service.index.dataset_version,
      engine: 'hormonerx-backend-ts/0.3.0',
      extractionModel: EXTRACTOR_VERSION,
      mode: 'demo-deterministic',
      evalTimestamp: new Date().toISOString(),
      deterministic: true,
      runtimeEligibleRecords: service.index.runtimeEligibleIds(),
      pendingPhysicianSignOff: Object.entries(service.index.reports)
        .filter(([, rep]) => rep.eligible_via_pending_override)
        .map(([rid]) => rid),
    },
    layers: {},
  };
  if (layers.includes('text')) {
    output.layers.text = await runTextLayer(service);
  }
  if (layers.includes('streaming')) {
    output.layers.streaming = await runStreamingLayer(service);
  }
  if (layers.includes('audio')) {
    output.layers.audio = runAudioLayer();
  }

  const resultsPath = path.join(BACKEND_DIR, 'data', 'benchmark_results.json');
  writeFileSync(resultsPath, JSON.stringify(output, null, 2) + '\n');

  const uiCopy = path.join(REPO_DIR, 'src', 'data', 'streaming_benchmark_results.json');
  writeFileSync(uiCopy, JSON.stringify(output, null, 2) + '\n');

  for (const [name, layer] of Object.entries(output.layers) as Array<[string, any]>) {
    if (layer.status === 'SKIPPED') {
      console.log(`[${name}] SKIPPED: ${layer.reason}`);
      continue;
    }
    console.log(`[${name}] cases=${layer.caseCount} metrics=${JSON.stringify(layer.metrics)}`);
    const failing = (layer.cases ?? []).filter((c: any) => !(c.pass ?? true)).map((c: any) => c.id);
    if (failing.length) {
      console.log(`[${name}] FAILING: ${JSON.stringify(failing)}`);
    }
  }
  console.log(`Results written to ${resultsPath} and ${uiCopy}`);
}

const isMain =
  process.argv[1] !== undefined && import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;

if (isMain) {
  const layerArgs: string[] = [];
  for (let i = 2; i < process.argv.length; i++) {
    if (process.argv[i] === '--layer' && process.argv[i + 1]) {
      layerArgs.push(process.argv[++i]);
    }
  }
  let selected = layerArgs.length ? layerArgs : ['all'];
  if (selected.includes('all')) {
    selected = ['text', 'streaming', 'audio'];
  }
  main(selected).catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: backend/app/benchmark.py
"""Reproducible benchmark runner (spec §25, §26).

Layers:
  text      — Layer A: standalone snippets, one finalized turn each.
  streaming — Layer B: event sequences with an expected state after every event.
  audio     — Layer C: requires recorded synthetic audio + live transcription;
              reported as skipped (never fabricated) when prerequisites are absent.

Gold labels are read-only inputs. Results are written to
backend/data/benchmark_results.json and a UI copy of the streaming summary to
src/data/streaming_benchmark_results.json. Real numbers only.
"""
from __future__ import annotations

import argparse
import asyncio
import json
import statistics
import sys
from datetime import datetime, timezone
from pathlib import Path

from .config import get_settings
from .deterministic_extractor import EXTRACTOR_VERSION, DeterministicExtractor
from .encounter_service import DuplicateEventError, EncounterService
from .evidence_index import EvidenceIndex
from .models import Speaker, new_id
from .warning_engine import NO_MATCH_PRIMARY, NO_MATCH_SECONDARY

BACKEND_DIR = Path(__file__).resolve().parent.parent
REPO_DIR = BACKEND_DIR.parent

ALLOWED_MESSAGES = {NO_MATCH_PRIMARY, NO_MATCH_SECONDARY}


def _round(x: float) -> float:
    return round(x, 4)


def _safe_div(a: float, b: float) -> float:
    return 0.0 if b == 0 else a / b


def build_service() -> EncounterService:
    settings = get_settings()
    index = EvidenceIndex(
        settings.evidence_path,
        settings.synonym_path,
        strict=True,
        allow_pending_verification=settings.evidence_allow_pending_verification,
    )
    return EncounterService(settings, index, DeterministicExtractor(index))


# ---------------------------------------------------------------------------
# Layer A — text
# ---------------------------------------------------------------------------

async def run_text_layer(service: EncounterService) -> dict:
    cases = json.loads((BACKEND_DIR / "data" / "benchmark_cases.json").read_text())["cases"]
    tp = fp = fn = tn = 0
    retrieval_correct = 0
    abst_expected = abst_correct = 0
    neg_total = neg_correct = hist_total = hist_correct = 0
    expected_positive = cited_correct = 0
    unsupported = 0
    status_total = status_correct = 0
    pass_count = 0
    results = []

    for case in cases:
        runtime = service.create_encounter()
        snapshot = await service.process_final_turn(
            runtime, event_id=new_id("evt"), text=case["input"], speaker=Speaker.DOCTOR
        )
        actual_state = snapshot.result_state.value
        active = snapshot.active_warnings()
        actual_record = active[0].evidence_record_id if len(active) == 1 else (None if not active else ",".join(sorted(w.evidence_record_id for w in active)))
        state_ok = actual_state == case["expectedResultState"]
        record_ok = actual_record == case["expectedEvidenceRecordId"]
        passed = state_ok and record_ok
        pass_count += passed

        exp_pos = case["expectedResultState"] == "EVIDENCE_FOUND"
        act_pos = actual_state == "EVIDENCE_FOUND"
        if exp_pos:
            expected_positive += 1
        if act_pos and exp_pos and record_ok:
            tp += 1
        elif act_pos and (not exp_pos or not record_ok):
            fp += 1
        if exp_pos and (not act_pos or not record_ok):
            fn += 1
        if not act_pos and not exp_pos:
            tn += 1
        if record_ok:
            retrieval_correct += 1
        if case["expectedAbstention"]:
            abst_expected += 1
            if not act_pos:
                abst_correct += 1
        if case["category"] == "explicit_negation":
            neg_total += 1
            neg_correct += state_ok
        if case["category"] == "historical_use":
            hist_total += 1
            hist_correct += state_ok
        if act_pos:
            record = service.index.get_record(active[0].evidence_record_id)
            if record_ok and record["sourceUrl"].strip() and record["sourceSection"].strip():
                cited_correct += 1
        for message in snapshot.messages:
            if message not in ALLOWED_MESSAGES:
                unsupported += 1
        # Medication status accuracy for the interacting medication, when labeled.
        expected_status = case.get("expectedMedicationStatus")
        if expected_status and case.get("expectedMedication"):
            status_total += 1
            interacting = [
                a for a in snapshot.assertions
                if a.category.value == "other_medication" and a.concept_id in service.index.ontology.medication_concepts
            ]
            if any(a.status.value == expected_status for a in interacting):
                status_correct += 1

        results.append(
            {
                "id": case["id"],
                "category": case["category"],
                "input": case["input"],
                "expectedResultState": case["expectedResultState"],
                "actualResultState": actual_state,
                "expectedEvidenceRecordId": case["expectedEvidenceRecordId"],
                "actualEvidenceRecordId": actual_record,
                "pass": passed,
                "lookupReason": snapshot.lookup_reason,
            }
        )

    precision = _safe_div(tp, tp + fp)
    recall = _safe_div(tp, tp + fn)
    return {
        "layer": "A-text",
        "caseCount": len(cases),
        "metrics": {
            "passRate": _round(_safe_div(pass_count, len(cases))),
            "triggerPrecision": _round(precision),
            "triggerRecall": _round(recall),
            "triggerF1": _round(_safe_div(2 * precision * recall, precision + recall)),
            "retrievalAccuracy": _round(_safe_div(retrieval_correct, len(cases))),
            "correctAbstentionRate": _round(_safe_div(abst_correct, abst_expected)),
            "negationAccuracy": _round(_safe_div(neg_correct, neg_total)),
            "historicalContextAccuracy": _round(_safe_div(hist_correct, hist_total)),
            "citationCoverage": _round(_safe_div(cited_correct, expected_positive)),
            "medicationStatusAccuracy": _round(_safe_div(status_correct, status_total)),
            "falsePositiveCount": fp,
            "unsupportedClaimCount": unsupported,
        },
        "confusion": {"truePositive": tp, "falsePositive": fp, "falseNegative": fn, "trueNegative": tn},
        "counts": {"passCount": pass_count, "total": len(cases)},
        "cases": results,
    }


# ---------------------------------------------------------------------------
# Layer B — streaming
# ---------------------------------------------------------------------------

async def run_streaming_layer(service: EncounterService) -> dict:
    data = json.loads((BACKEND_DIR / "data" / "streaming_benchmark_cases.json").read_text())
    cases = data["cases"]
    event_total = event_correct = 0
    final_correct = 0
    retrieval_events = retrieval_correct = 0
    retraction_expected = retraction_correct = 0
    premature_warnings = 0
    duplicate_warnings = 0
    citation_events = citation_correct = 0
    unsupported = 0
    retraction_latency_turns: list[int] = []
    case_results = []

    for case in cases:
        runtime = service.create_encounter()
        proposal_by_sequence: dict[int, str] = {}
        event_results = []
        warning_created_at_event: dict[str, int] = {}
        for event_index, event in enumerate(case["events"]):
            etype = event["type"]
            try:
                if etype == "transcript.final":
                    await service.process_final_turn(
                        runtime,
                        event_id=event.get("event_id") or new_id("evt"),
                        text=event["text"],
                        speaker=Speaker(event["speaker"]),
                        sequence=event["sequence"],
                    )
                elif etype == "prescription.proposed":
                    snapshot = await service.propose_prescription(
                        runtime, event_id=new_id("evt"), surface_text=event["medication_surface_text"]
                    )
                    proposal_by_sequence[event["sequence"]] = snapshot.proposals[-1].proposal_id
                elif etype == "prescription.cancelled":
                    ref = event["cancelsEventOfSequence"]
                    await service.cancel_prescription(
                        runtime, event_id=new_id("evt"), proposal_id=proposal_by_sequence[ref]
                    )
            except DuplicateEventError:
                if not event.get("isDuplicateReplay"):
                    raise

            snapshot = runtime.snapshot
            active = snapshot.active_warnings()
            active_records = sorted({w.evidence_record_id for w in active})
            state_ok = snapshot.result_state.value in event["expectedResultStates"]
            records_ok = active_records == sorted(event["expectedEvidenceRecordIds"])
            count_ok = len(active) == event["expectedActiveWarningCount"]
            retraction_ok = True
            if event.get("expectRetractionWithReason"):
                retraction_expected += 1
                retracted = [w for w in snapshot.warnings if w.state == "retracted"]
                retraction_ok = bool(retracted and all(w.retraction_reason for w in retracted))
                retraction_correct += retraction_ok
                for w in retracted:
                    if w.warning_id in warning_created_at_event:
                        retraction_latency_turns.append(event_index - warning_created_at_event.pop(w.warning_id))
            context_ok = True
            if event.get("expectedWarningContext"):
                context_ok = bool(active) and all(w.context.value == event["expectedWarningContext"] for w in active)
            for w in active:
                warning_created_at_event.setdefault(w.warning_id, event_index)
            if event["expectedEvidenceRecordIds"]:
                retrieval_events += 1
                retrieval_correct += records_ok
                citation_events += len(active)
                for w in active:
                    record = service.index.get_record(w.evidence_record_id)
                    citation_correct += bool(record["sourceUrl"].strip() and record["sourceSection"].strip())
            if event["expectedActiveWarningCount"] == 0 and active:
                premature_warnings += len(active)
            if len(active) > event["expectedActiveWarningCount"]:
                duplicate_warnings += len(active) - event["expectedActiveWarningCount"]
            for message in snapshot.messages:
                if message not in ALLOWED_MESSAGES:
                    unsupported += 1

            ok = state_ok and records_ok and count_ok and retraction_ok and context_ok
            event_total += 1
            event_correct += ok
            event_results.append(
                {
                    "sequence": event.get("sequence"),
                    "type": etype,
                    "expectedResultStates": event["expectedResultStates"],
                    "actualResultState": snapshot.result_state.value,
                    "expectedEvidenceRecordIds": event["expectedEvidenceRecordIds"],
                    "actualEvidenceRecordIds": active_records,
                    "pass": ok,
                }
            )
        final_ok = event_results[-1]["pass"]
        final_correct += final_ok
        case_results.append(
            {
                "id": case["id"],
                "category": case["category"],
                "pass": all(e["pass"] for e in event_results),
                "events": event_results,
            }
        )

    latencies = [
        latency.total_ms
        for case_runtime in service.encounters.values()
        for latency in case_runtime.latencies
    ]
    latency_summary = None
    if latencies:
        latencies_sorted = sorted(latencies)
        latency_summary = {
            "samples": len(latencies_sorted),
            "medianMs": _round(statistics.median(latencies_sorted)),
            "p90Ms": _round(latencies_sorted[int(0.9 * (len(latencies_sorted) - 1))]),
            "maxMs": _round(latencies_sorted[-1]),
            "note": "Backend pipeline only (final transcript received -> result computed), deterministic extractor, no network. Live-mode model latency is not included and must be measured separately.",
        }

    return {
        "layer": "B-streaming",
        "caseCount": len(cases),
        "metrics": {
            "casePassRate": _round(_safe_div(sum(c["pass"] for c in case_results), len(cases))),
            "perEventStateAccuracy": _round(_safe_div(event_correct, event_total)),
            "finalStateAccuracy": _round(_safe_div(final_correct, len(cases))),
            "retrievalAccuracy": _round(_safe_div(retrieval_correct, retrieval_events)),
            "warningRetractionAccuracy": _round(_safe_div(retraction_correct, retraction_expected)),
            "meanRetractionLatencyEvents": _round(_safe_div(sum(retraction_latency_turns), len(retraction_latency_turns))) if retraction_latency_turns else 0.0,
            "prematureWarningCount": premature_warnings,
            "duplicateWarningCount": duplicate_warnings,
            "citationCoverage": _round(_safe_div(citation_correct, citation_events)),
            "unsupportedClaimCount": unsupported,
        },
        "counts": {"events": event_total, "eventsCorrect": event_correct},
        "processingLatency": latency_summary,
        "cases": case_results,
    }


# ---------------------------------------------------------------------------
# Layer C — audio (prerequisite-gated; never fabricated)
# ---------------------------------------------------------------------------

def run_audio_layer(service: EncounterService) -> dict:
    manifest = json.loads((BACKEND_DIR / "data" / "audio_benchmark_manifest.json").read_text())
    settings = get_settings()
    missing_files = [
        c["file"] for c in manifest["cases"] if not (BACKEND_DIR / "data" / c["file"]).exists()
    ]
    reasons = []
    if missing_files:
        reasons.append(f"{len(missing_files)} audio files not recorded yet")
    if not settings.openai_api_key:
        reasons.append("no OPENAI_API_KEY for live transcription")
    if reasons:
        return {
            "layer": "C-audio",
            "status": "SKIPPED",
            "reason": "; ".join(reasons),
            "caseCount": len(manifest["cases"]),
            "note": "Results are only reported for runs that actually happened. Record the manifest audio and set OPENAI_API_KEY to execute this layer.",
        }
    return {
        "layer": "C-audio",
        "status": "NOT_IMPLEMENTED_IN_THIS_RUNNER_VERSION",
        "note": "Transcribe each file through the realtime pipeline and evaluate against the manifest gold labels.",
    }


# ---------------------------------------------------------------------------

async def main(layers: list[str]) -> None:
    service = build_service()
    output: dict = {
        "metadata": {
            "benchmarkVersion": "0.2.0",
            "evidenceVersion": service.index.dataset_version,
            "engine": "hormonerx-backend/0.2.0",
            "extractionModel": EXTRACTOR_VERSION,
            "mode": "demo-deterministic",
            "evalTimestamp": datetime.now(timezone.utc).isoformat(),
            "deterministic": True,
            "runtimeEligibleRecords": service.index.runtime_eligible_ids(),
            "pendingPhysicianSignOff": [
                rid for rid, rep in service.index.reports.items() if rep.eligible_via_pending_override
            ],
        },
        "layers": {},
    }
    if "text" in layers:
        output["layers"]["text"] = await run_text_layer(service)
    if "streaming" in layers:
        output["layers"]["streaming"] = await run_streaming_layer(service)
    if "audio" in layers:
        output["layers"]["audio"] = run_audio_layer(service)

    results_path = BACKEND_DIR / "data" / "benchmark_results.json"
    results_path.write_text(json.dumps(output, indent=2, ensure_ascii=False) + "\n")

    ui_copy = REPO_DIR / "src" / "data" / "streaming_benchmark_results.json"
    ui_copy.write_text(json.dumps(output, indent=2, ensure_ascii=False) + "\n")

    for name, layer in output["layers"].items():
        if layer.get("status") == "SKIPPED":
            print(f"[{name}] SKIPPED: {layer['reason']}")
            continue
        print(f"[{name}] cases={layer['caseCount']} metrics={json.dumps(layer['metrics'])}")
        failing = [c["id"] for c in layer.get("cases", []) if not c.get("pass", True)]
        if failing:
            print(f"[{name}] FAILING: {failing}")
    print(f"Results written to {results_path} and {ui_copy}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--layer", action="append", choices=["text", "streaming", "audio", "all"], default=None)
    args = parser.parse_args()
    selected = args.layer or ["all"]
    if "all" in selected:
        selected = ["text", "streaming", "audio"]
    asyncio.run(main(selected))
